desci-server/src/services/data/externalCidProcessing.ts

```diff
@@ -0,0 +1,213 @@
+import { User, Node, Prisma } from '@prisma/client';
+
+import prisma from 'client';
+import { persistManifest } from 'controllers/data/utils';
+import parentLogger from 'logger';
+import { ensureUniquePathsDraftTree } from 'services/draftTrees';
+import { GetExternalSizeAndTypeResult, convertToCidV1, getExternalCidSizeAndType, pubRecursiveLs } from 'services/ipfs';
+import { FirstNestingComponent, addComponentsToManifest, getTreeAndFill } from 'utils/driveUtils';
+
+import { getManifestFromNode, updateDataReferences } from './processing';
+import { createIpfsUnresolvableError, createManifestPersistFailError, createUnhandledError } from './processingErrors';
+
+const logger = parentLogger.child({
+  module: 'Services::ExternalCidProcessing',
+});
+
+interface ProcessExternalCidDataToIpfsParams {
+  // files: any[];
+  externalCids: any;
+  user: User;
+  node: Node;
+  /**
+   * @type {string} path to the directory to be updated
+   */
+  contextPath: string;
+  componentType?: ResearchObjectComponentType;
+  componentSubtype?: ResearchObjectComponentSubtypes;
+}
+
+/**
+ * Processes external-url file uploads, currently .pdf URLs and github code repo's
+ */
+export async function processExternalCidDataToIpfs({
+  externalCids,
+  user,
+  node,
+  contextPath,
+  componentType,
+  componentSubtype,
+}: ProcessExternalCidDataToIpfsParams) {
+  try {
+    // debugger;
+    /**
+     * Prepare the CIDs for addition, see if they're resolvable and get their sizes and types
+     */
+    const cidTypesSizes: Record<string, GetExternalSizeAndTypeResult> = {};
+    if (externalCids && externalCids.length) {
+      try {
+        externalCids = externalCids.map((extCid) => ({ ...extCid, cid: convertToCidV1(extCid.cid) }));
+        for (const extCid of externalCids) {
+          const { isDirectory, size } = await getExternalCidSizeAndType(extCid.cid);
+          if (size !== undefined && isDirectory !== undefined) {
+            cidTypesSizes[extCid.cid] = { size, isDirectory };
+          } else {
+            throw new Error(`Failed to get size and type of external CID: ${extCid}`);
+          }
+        }
+      } catch (e: any) {
+        logger.warn(`[UPDATE DAG] External CID Method: ${e}`);
+        throw createIpfsUnresolvableError(`Failed to resolve external CID`);
+      }
+    }
+
+    /**
+     * Ensure no collisions
+     */
+    const cidFileNames = externalCids.map((extCid) => extCid.name);
+    await ensureUniquePathsDraftTree({
+      nodeId: node.id,
+      contextPath,
+      externalUrlFilePaths: cidFileNames,
+    });
+
+    /**
+     * Discover the entire tree if it's a DAG, prepare entries for the draft tree
+     */
+    const externalCidMap = {};
+    let entriesDiscovered = [];
+    // const externalDagsToPin = [];
+    if (externalCids?.length && Object.keys(cidTypesSizes)?.length) {
+      entriesDiscovered = [];
+      for await (const extCid of externalCids) {
+        const { size, isDirectory } = cidTypesSizes[extCid.cid];
```

Comment:
Does this assume the CID is necessarily unixFS? Curious what'd happen if it's some other file structure.